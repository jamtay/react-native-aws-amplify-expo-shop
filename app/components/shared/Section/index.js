import React, {useState} from 'react';

import {StyleSheet, Text, TouchableHighlight, View} from 'react-native';
import {Icon} from 'native-base';
import Modal from '../Modal';
import {COLOURS} from '../../../styles/colours';
import {BUTTONS} from '../../../styles/button';

/**
 * A section used with a border bottom to display information for a single store
 * @param title The title of the section
 * @param textLabel The text label prompt for data entry in the modal
 * @param onDataSubmit Function called when data is entered
 * @param keyboardType Type of keyboard for text input e.g "number-pad" for the modal
 * @param multiItemEntry Is the data entry multiple lines or a single line
 * @param children React elements to display underneath the title
 */
const Section = ({
  title,
  children,
  textLabel,
  onDataSubmit,
  keyboardType,
  multiItemEntry,
}) => {
  const [isModalVisible, setModalVisible] = useState(false);
  const onNewRecordingButtonClick = () => {
    setModalVisible(true);
  };

  return (
    <View style={styles.border}>
      <View style={styles.section} testID="section-wrapper">
        <View style={[styles.textWrapper]}>
          <Text style={styles.titleText}>{title}</Text>
        </View>
        <View style={styles.buttonWrapper}>
          <TouchableHighlight
            testID="section-button"
            onPress={onNewRecordingButtonClick}
            activeOpacity={BUTTONS.IMAGE_CLICK_OPACITY}
            underlayColor={BUTTONS.CLICK_COLOUR}>
            <Icon
              name="add-circle-outline"
              type="MaterialIcons"
              style={styles.button}
            />
          </TouchableHighlight>
        </View>
        <Modal
          isVisible={isModalVisible}
          textLabel={textLabel}
          onDismiss={() => setModalVisible(false)}
          onDataSubmit={onDataSubmit}
          keyboardType={keyboardType}
          multiItemEntry={multiItemEntry}
        />
      </View>
      {children}
    </View>
  );
};

export default Section;

const styles = StyleSheet.create({
  titleText: {
    fontSize: 20,
    fontWeight: '700',
    marginTop: 20,
    paddingHorizontal: 20,
  },
  textWrapper: {
    flex: 5,
    marginBottom: 5,
  },
  buttonWrapper: {
    flex: 1,
    top: 6,
  },
  section: {
    paddingBottom: 10,
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingLeft: -10,
    alignItems: 'center',
  },
  border: {
    borderBottomWidth: 1,
    borderBottomColor: COLOURS.DARK_PINK,
    marginHorizontal: 20,
    paddingVertical: 10,
  },
  button: {
    color: COLOURS.DARK_PINK,
    marginRight: 10,
  },
});
