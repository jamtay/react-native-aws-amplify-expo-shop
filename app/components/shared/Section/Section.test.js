import React from 'react';

import renderer from 'react-test-renderer';
import {fireEvent, render} from 'react-native-testing-library';
import {Text} from 'react-native';

import Section from './index';

describe('Section component tests', () => {
  const onDataSubmitMock = jest.fn();
  const childrenText = 'childrenText';
  const props = {
    title: 'SOME_TITLE',
    children: <Text>{childrenText}</Text>,
    textLabel: 'SOME_TEXT_LABEL',
    onDataSubmit: onDataSubmitMock,
    keyboardType: 'number-pad',
    multiItemEntry: false,
  };

  describe('snapshot test', () => {
    it('should match stored snapshot', () => {
      const tree = renderer.create(<Section {...props} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('render with properties', () => {
    it('should render the supplied title', async () => {
      const {getByText} = render(<Section {...props} />);
      const title = await getByText(props.title);
      expect(title).toBeTruthy();
    });

    it('should render the supplied children inside the component', async () => {
      const {getByText} = render(<Section {...props} />);
      const childElement = await getByText(childrenText);
      expect(childElement).toBeTruthy();
    });

    it('should display modal when button is clicked', async () => {
      const {getByTestId} = render(<Section {...props} />);
      const openModalButton = await getByTestId('section-button');
      const sectionModal = await getByTestId('section-wrapper');
      const modal = await getByTestId('modal-component');

      expect(modal.props.children.props.isVisible).toEqual(false);
      expect(sectionModal.props.children[2].props.isVisible).toEqual(false);
      fireEvent.press(openModalButton);
      expect(modal.props.children.props.isVisible).toEqual(true);
      expect(sectionModal.props.children[2].props.isVisible).toEqual(true);
    });

    it('should use supplied textLabel, keyboardType, multiItemEntry within the modal', async () => {
      const {getByTestId, getByText} = render(<Section {...props} />);
      const textLabel = getByText(props.textLabel);
      const sectionModal = await getByTestId('section-wrapper');

      expect(textLabel).toBeTruthy();
      expect(sectionModal.props.children[2].props.textLabel).toEqual(
        props.textLabel,
      );
      expect(sectionModal.props.children[2].props.keyboardType).toEqual(
        props.keyboardType,
      );
      expect(sectionModal.props.children[2].props.multiItemEntry).toEqual(
        props.multiItemEntry,
      );
    });

    it('should call onDataSubmit when adding a new entry to the modal', async () => {
      const {getByTestId, getByText} = render(<Section {...props} />);
      const openModalButton = await getByTestId('section-button');
      const submitButton = await getByText('Submit');
      const input = await getByTestId('single-input');

      const dataInput = 'some text to input';
      fireEvent.press(openModalButton);
      fireEvent.changeText(input, dataInput);
      fireEvent.press(submitButton);

      expect(onDataSubmitMock).toBeCalledTimes(1);
      expect(onDataSubmitMock).toBeCalledWith([dataInput]);
    });
  });
});
