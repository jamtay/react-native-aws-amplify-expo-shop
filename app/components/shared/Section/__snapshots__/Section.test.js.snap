// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section component tests snapshot test should match stored snapshot 1`] = `
<View
  style={
    Object {
      "borderBottomColor": "violet",
      "borderBottomWidth": 1,
      "marginHorizontal": 20,
      "paddingVertical": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 10,
        "paddingLeft": -10,
      }
    }
    testID="section-wrapper"
  >
    <View
      style={
        Array [
          Object {
            "flex": 5,
            "marginBottom": 5,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "fontWeight": "700",
            "marginTop": 20,
            "paddingHorizontal": 20,
          }
        }
      >
        SOME_TITLE
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "top": 6,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={null}
        testID="section-button"
      >
        <Text />
      </View>
    </View>
    <View
      testID="modal-component"
    >
      <Modal
        animationType="none"
        hardwareAccelerated={false}
        hideModalContentWhileAnimating={false}
        onDismiss={[Function]}
        onModalHide={[Function]}
        onModalWillHide={[Function]}
        onModalWillShow={[Function]}
        onRequestClose={[Function]}
        onSwipeComplete={[Function]}
        scrollHorizontal={false}
        scrollOffset={0}
        scrollOffsetMax={0}
        scrollTo={null}
        supportedOrientations={
          Array [
            "portrait",
            "landscape",
          ]
        }
        swipeDirection={
          Array [
            "down",
            "up",
            "left",
            "right",
          ]
        }
        swipeThreshold={100}
        transparent={true}
        visible={false}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "black",
              "bottom": 0,
              "height": 1334,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 750,
            }
          }
        />
        <View
          hideModalContentWhileAnimating={false}
          onDismiss={[Function]}
          onModalHide={[Function]}
          onModalWillHide={[Function]}
          onModalWillShow={[Function]}
          onMoveShouldSetResponder={[Function]}
          onMoveShouldSetResponderCapture={[Function]}
          onResponderEnd={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderReject={[Function]}
          onResponderRelease={[Function]}
          onResponderStart={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onStartShouldSetResponderCapture={[Function]}
          onSwipeComplete={[Function]}
          pointerEvents="box-none"
          scrollHorizontal={false}
          scrollOffset={0}
          scrollOffsetMax={0}
          scrollTo={null}
          style={
            Object {
              "flex": 1,
              "justifyContent": "center",
              "left": 0,
              "margin": 37.5,
              "top": 0,
              "transform": Array [
                Object {
                  "translateY": 0,
                },
              ],
            }
          }
          supportedOrientations={
            Array [
              "portrait",
              "landscape",
            ]
          }
          swipeDirection={
            Array [
              "down",
              "up",
              "left",
              "right",
            ]
          }
          swipeThreshold={100}
        >
          <View
            style={
              Object {
                "alignItems": "flex-end",
                "right": 5,
                "top": 30,
                "zIndex": 999,
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "top": 8,
                }
              }
              testID="dismiss-button"
            >
              <Text />
            </View>
          </View>
          <View
            style={
              Object {
                "backgroundColor": "white",
                "borderColor": "violet",
                "borderRadius": 8,
                "borderWidth": 4,
                "padding": 22,
                "paddingBottom": 35,
              }
            }
          >
            <View
              accessible={true}
              focusable={false}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderBottomWidth": 1,
                  "borderColor": "#D9D5DC",
                  "borderLeftWidth": 0,
                  "borderRadius": undefined,
                  "borderRightWidth": 0,
                  "borderTopWidth": 0,
                  "borderWidth": undefined,
                  "flexDirection": "row",
                  "marginBottom": 10,
                  "marginLeft": 2,
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "left": 0,
                    "opacity": 1,
                    "paddingBottom": undefined,
                    "paddingTop": undefined,
                    "position": "absolute",
                    "right": 0,
                    "top": 18,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#575757",
                        "fontSize": 17,
                        "paddingRight": 5,
                        "paddingTop": 5,
                      },
                      Object {
                        "left": 5,
                        "paddingTop": 10,
                        "textAlign": "left",
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontSize": 17,
                        },
                        Object {
                          "left": 5,
                          "paddingTop": 10,
                          "textAlign": "left",
                        },
                      ]
                    }
                  >
                    SOME_TEXT_LABEL
                  </Text>
                </Text>
              </View>
              <TextInput
                allowFontScaling={true}
                editable={true}
                keyboardType="number-pad"
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                placeholder={null}
                placeholderTextColor="#575757"
                rejectResponderTermination={true}
                style={
                  Object {
                    "color": "#000",
                    "flex": 1,
                    "fontSize": 17,
                    "height": 50,
                    "paddingBottom": 7,
                    "paddingLeft": 5,
                    "paddingRight": 5,
                    "paddingTop": 3,
                    "top": 8,
                  }
                }
                testID="single-input"
                textAlign="center"
                underlineColorAndroid="rgba(0,0,0,0)"
                value=""
              />
            </View>
            <View
              accessibilityRole="button"
              accessibilityStates={Array []}
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {},
                  ]
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#007AFF",
                        "fontSize": 18,
                        "padding": 8,
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  Submit
                </Text>
              </View>
            </View>
          </View>
        </View>
      </Modal>
    </View>
  </View>
  <Text>
    childrenText
  </Text>
</View>
`;
